Lib/urllib/request.py

```diff
@@ -1487,21 +1463,32 @@ def open_local_file(self, req):
             headers = email.message_from_string(
                 'Content-type: %s\nContent-length: %d\nLast-modified: %s\n' %
                 (mtype or 'text/plain', size, modified))
-            if host:
-                host, port = _splitport(host)
-            if not host or \
-                (not port and _safe_gethostbyname(host) in self.get_names()):
-                origurl = 'file:' + pathname2url(localfile)
-                return addinfourl(open(localfile, 'rb'), headers, origurl)
+            origurl = 'file:' + pathname2url(localfile)
+            return addinfourl(open(localfile, 'rb'), headers, origurl)
         except OSError as exp:
             raise URLError(exp, exp.filename)
-        raise URLError('file not on local host')
 
-def _safe_gethostbyname(host):
+    file_open = open_local_file
+
+_local_addresses = None
+
+def _is_local_authority(authority):
+    global _local_addresses
+
+    if not authority or authority == 'localhost':
+        return True
     try:
-        return socket.gethostbyname(host)
-    except socket.gaierror:
-        return None
+        address = socket.gethostbyname(authority)
+    except (socket.gaierror, AttributeError):
+        return False
+    if _local_addresses is None:
+        try:
+            _local_addresses = tuple(
+                socket.gethostbyname_ex('localhost')[2] +
+                socket.gethostbyname_ex(socket.gethostname())[2])
+        except socket.gaierror:
+            _local_addresses = (socket.gethostbyname('localhost'),)
+    return address in _local_addresses
```

Comment:
if doing membership testing, is it worth making `_local_addresses` a (frozen)set?